[Proposal] public Actions.empty

guizmaii opened the pull request:
# Pull Request Checklist

* [x] Have you read through the [contributor guidelines](https://github.com/lagom/lagom/tree/master/CONTRIBUTING.md)?
* [x] Have you signed the [Lightbend CLA](https://www.lightbend.com/contribute/cla)?
* [x] Have you [squashed your commits](https://github.com/lagom/lagom/tree/master/WorkingWithGit.md)?
* [x] Have you added copyright headers to new files?
* [x] Have you updated the documentation?
* [x] Have you added tests for any changed functionality?

## Fixes

Nothing

## Purpose

IMHO, `Actions.empty` function should be public because when we want to create an empty actions set, `Actions()` doesn't look like "scala like" and is not explicit enough.

When we want to create a empty List in Scala we do `List.empty`, so why not for `Actions` ?

## Background Context

Explained above

## References

None

rcavalcanti commented:
I'm for it! 👍

ignasi35 commented:
Hmmm, I have undecided opinions on this one. 

On one side I agree with the proposal.
On the other I think that an empty behavior in a PersistentEntity is unfinished business. It's a behavior you can't escape and where every command causes an `UnhandledCommandException`. So, exposing `empty` is not such a good idea.

I understand the purpose of the proposal is to make this more idiomatic.Maybe an alternative would be to provide fluent methods on the `object`:

```scala
  object Actions {
    private val empty = new Actions(PartialFunction.empty, Map.empty)
    def apply(): Actions = empty

    def onCommand[C <: Command with PersistentEntity.ReplyType[Reply]: ClassTag, Reply](
      handler: PartialFunction[(Command, CommandContext[Reply], State), Persist]
    ): Actions = {
      empty.onCommand(handler) 
    }

    def onReadOnlyCommand[C <: Command with PersistentEntity.ReplyType[Reply]: ClassTag, Reply](
       handler: PartialFunction[(Command, ReadOnlyCommandContext[Reply], State), Unit]
    ): Actions = {
      empty.onReadOnlyCommand(handler)
    }

    def onEvent(handler: EventHandler): Actions = empty.onEvent(handler)
    
  }

```
These methods are duplicates of those available in `class Actions` and allow the user to declare their actions using: (

```scala
 val b1: Behavior = Actions.onEvent(...)
```

directly, without using the `()` (parens) or `empty`.

Actually a user would even be allowed to:

```scala
 import Actions._ 
 val b1: Behavior = onEvent(...)
```
and I'm not sure this is a good idea. 🤔 

What do you think @rcavalcanti @guizmaii ?

rcavalcanti commented:
I think that in both cases (`Actions()` or `Actions.empty`) the intention is to use it as a place-holder to be filled later when the user decide to implement it. So, offering `Actions.empty` is not doing harm here.

rcavalcanti commented:
BTW, I like the idea to add extra methods in `Actions` companion object. 

I would add the three new methods plus make `empty` public.

guizmaii commented:
Hi,

I agree with the last comments of @rcavalcanti.
I need the `Action.empty` because, in some cases, I have some empty actions (see the example bellow)
The new functions you propose @ignasi35 seem great but how do you compose Actions:
```
Actions.onCommand {
  case ... => ...
  case ... => ...
} ++ Actions.onEvent {
  case ... => ...
  case ... => ...
}
```
???

IMHO (and I don't know if it's really possible), what we need here is that `type Behavior = State => Actions` in `PersistentEntity` become `type Behavior = PartialFunction[State, Action]` in order to be able to write:

```
Actions
  .onReadOnlyCommand {
    case (GetCar(), ctx, state) => ctx.reply(state.maybeCar.get)
  }
  .onCommand {
    case (UpdateCarName(name), ctx, state) =>
      ctx.thenPersist(CarNameUpdated(entityId = entityId, name = name)(_ => ctx.reply(Done))
    case (DeleteCar(), ctx, state) =>
      ctx.thenPersist(CarDeleted(entityId = entityId)(_ => ctx.reply(Done))
  }
  .onEvent {
    case (CarNameUpdated(name), state) => state.updateCarName(name)
    case (CarDeleted(_), state) => state.deleteCar()
  }
```
instead of:
```
Actions()
  .onReadOnlyCommand[GetCar, Car] {
    case (GetCar(), ctx, state) => ctx.reply(state.maybeCar.get)
  }
  .onCommand[UpdateCarName, Done] {
    case (UpdateCarName(name), ctx, state) =>
      ctx.thenPersist(CarNameUpdated(entityId = entityId, name = name)(_ => ctx.reply(Done))
  }
  .onCommand[DeleteCar, Done] {
    case (DeleteCar(), ctx, state) =>
      ctx.thenPersist(CarDeleted(entityId = entityId)(_ => ctx.reply(Done))
  }
  .onEvent {
    case (CarNameUpdated(name), state) => state.updateCarName(name)
    case (CarDeleted(_), state) => state.deleteCar()
  }
```

A complete exemple of things we do to understand why I say that :

```
final class CarEntity extends PersistentEntity {

  import CarCommands._
  import CarEvents._

  override type Command = CarCommand
  override type Event   = CarEvent
  override type State   = CarState

  override def initialState: CarState = CarState.empty

  override def behavior: Behavior =
    state =>
      state.currentStatus match {
        case CarStatus.EMPTY   => carNotCreatedActions
        case CarStatus.CREATED => carCreatedActions
        case CarStatus.DELETED => carDeletedActions
    }

  private def carNotCreatedActions: Actions =
    Actions()
      .onCommand[CreateCar, Done] {
        case (CreateCar(car), ctx, state) =>
          ctx.thenPersist(CarCreated(car = car))(_ => ctx.reply(Done))
      }
      .onEvent {
        case (CarCreated(car), _) => CarState(car)
      }

  private def carCreatedActions: Actions =
    Actions()
      .onReadOnlyCommand[GetCar, Car] {
        case (GetCar(), ctx, state) => ctx.reply(state.maybeCar.get)
      }
      .onCommand[UpdateCarName, Done] {
        case (UpdateCarName(name), ctx, state) =>
          ctx.thenPersist(CarNameUpdated(entityId = entityId, name = name)(_ => ctx.reply(Done))
      }
      .onCommand[DeleteCar, Done] {
        case (DeleteCar(), ctx, state) =>
          ctx.thenPersist(CarDeleted(entityId = entityId)(_ => ctx.reply(Done))
      }
      .onEvent {
        case (CarNameUpdated(name), state) => state.updateCarName(name)
        case (CarDeleted(_), state) => state.deleteCar()
      }

  private def carDeletedActions: Actions = Actions()
```

With the new functions `.empty` and the `type Behavior = PartialFunction[State, Action]`, it'll look better: 
```
final class CarEntity extends PersistentEntity {

  import CarCommands._
  import CarEvents._

  override type Command = CarCommand
  override type Event   = CarEvent
  override type State   = CarState

  override def initialState: CarState = CarState.empty

  override def behavior: Behavior =
    state =>
      state.currentStatus match {
        case CarStatus.EMPTY   => carNotCreatedActions
        case CarStatus.CREATED => carCreatedActions
        case CarStatus.DELETED => carDeletedActions
    }

  private def carNotCreatedActions: Actions =
    Actions
      .onCommand {
        case (CreateCar(car), ctx, state) =>
          ctx.thenPersist(CarCreated(car = car))(_ => ctx.reply(Done))
      }
      .onEvent {
        case (CarCreated(car), _) => CarState(car)
      }

  private def carCreatedActions: Actions =
    Actions
      .onReadOnlyCommand {
        case (GetCar(), ctx, state) => ctx.reply(state.maybeCar.get)
      }
      .onCommand {
        case (UpdateCarName(name), ctx, state) =>
          ctx.thenPersist(CarNameUpdated(entityId = entityId, name = name)(_ => ctx.reply(Done))
        case (DeleteCar(), ctx, state) =>
          ctx.thenPersist(CarDeleted(entityId = entityId)(_ => ctx.reply(Done))
      }
      .onEvent {
        case (CarNameUpdated(name), state) => state.updateCarName(name)
        case (CarDeleted(_), state) => state.deleteCar()
      }

  private def carDeletedActions: Actions = Actions.empty
```

ignasi35 commented:
> The new functions you propose @ignasi35 seem great but how do you compose Actions:

```scala
Actions.onCommand {
  case ... => ...
  case ... => ...
} ++ Actions.onEvent {
  case ... => ...
  case ... => ...
}
```

@guizmaii My idea was to add convenience methods to the `object Actions` but keep all methods in `class Actions`.  In `class Actions` there already is a `orElse` which is the recommended way to compose Actions. So the code above (assuming the changes I suggested in the `object Actions`) would be:

val defaultActions = Actions.onCommand {
  case ... => ...
  case ... => ...
} 
val otherActions = defaultActions orElse Actions.onEvent {
  case ... => ...
  case ... => ...
}

In my original comment I purposedly skipped `orElse` when creating the convenience methods in the `object Actions` because it doesn't make sense to compose `empty` with something else.

ignasi35 commented:
```scala
  private def carDeletedActions: Actions = Actions.empty
```

Hmmm, in my previous comment I considered this a smell but the more I look at it the more I see this being a common use-case.

What I still don't like is that Behaviour causing an `UnhandledCommandException` on every case. I think it'd be good to have a method to throw a more domain-friendly exception instead of `UnhandledCommandException`. In your particular case where there's no commandHandler required because the car is _deleted_ I think the default behavior could be something like:

```scala
  // pseudoCode
  private def carDeletedActions: Actions = Actions.defaultTo { case _ => throw new NotFound }
```

ignasi35 commented:
PS: it'd be awesome if, on top of the `defaultTo` we restricted the default exception to subclasses of [`NoStacktrace`](https://www.scala-lang.org/api/2.11.8/#scala.util.control.NoStackTrace)

guizmaii commented:
I deleted the part of my previous where I talk about `type Behavior ...` because it was stupid. Sorry for that.

rcavalcanti commented:
This was actually a PR on 1.3.x branch.
And it's now forward ported to master ad092e5147b9772d4f681542e9c651fb18c21162

rcavalcanti commented:
Thanks @guizmaii !

guizmaii commented:
Thank you two 🙂

guizmaii commented:
Just as a last comment (that I hope you'll read even if the PR is merged 🙂) about what @ignasi35 said with the `Actions.defaultTo { case _ => throw new NotFound }`.

I'm not a big fan of throwing exceptions, I prefer types ! 
When exceptions are needed I largely prefer return for example a `Either[Exception, Result]`.

Is it an option to replace, in 1.4 for example, the exceptions thrown in entities by a "validator" type (Cats Validated type ? inspiration behind that idea: https://www.youtube.com/watch?v=0LX3i7uQ-wI).

It should looks something like that: `type Behavior = State => Validated[Exception, Result]`

What's your opinion on that @rcavalcanti @ignasi35 ?

rcavalcanti commented:
It would be very interesting to have something on those lines. I have myself play around in Fun.CQRS with the concept of having different types of effects returning from a command handler. There is even an open PR for it (https://github.com/strongtyped/fun-cqrs/pull/83).

However, this will be challenging with respect to the java API, but also from the point of view of java developers. 

On one of my previous projects, I had an application using `Validated[Errors, Result]`. This was the most powerful app we have built there (not all were using applicative). The domain was quite complex and a `Command` could fail for many different reasons and we were able to inform the user all the pieces that were wrong. 

I'm open for something like that, but that will require some extra efforts and lots of creativity on the java side. 

/cc @ignasi35 @TimMoore

ignasi35 commented:
Just a quick shallow response to manage expectetations.
`1.4.0` is too close for such a change.

This change is breaking the API so in a best case scenario it would be held until Lagom 2.0.0.

guizmaii commented:
There will be no `1.5.0` ?? 🙂

TimMoore commented:
Let's raise a separate issue to discuss this

guizmaii commented:
I will. Thanks
